blog/2020-10-01-hs-to-coq/2020-10-01-hs-to-coq.md

```diff
@@ -0,0 +1,1084 @@
+    split.
+    + reflexivity.
+    + apply models_MkQueue.
+      cbn.
+      rewrite app_nil_r.
+      reflexivity.
+  - cbn in Eqn_xs.
+    exists (MkQueue front back).
+    injection Eqn_xs.
+    intros Eqn_xs' Ey.
+    split.
+    + rewrite Ey. reflexivity.
+    + apply models_MkQueue. assumption.
+Qed.
+```
+</div>
+
+
+### Discussion
+
+What good is it to formally verify a library?
+
+The informal marketing slogan for formal verification is that "it guarantees
+the absence of bugs". What we actually proved is that the implementation
+conforms to a model, a specification. We can thus say that this proof
+guarantees that the implementation is at least as good as the specification.
+But how good is the specification?
+
+The obvious way to evaluate the specification is to read it, and convince
+oneself that, yes, that's what a queue should do. For that reason, it is
+desirable for the specification to be significantly simpler than the
+implementation.
+
+The specification can also be seen as an abstraction of the implementation,
+describing the features that are relevant to users of the implementation,
+while hiding other internal *implementation details*.
+A *rich* specification is one that allows meaningful properties to be proved
+about programs using a conforming implementation.
+
+At the very least, we should be able to use the specification to prove
+simple properties of toy programs, as a form of *unit testing* of the
+specification.
+
+For instance, the following unit test is a proof that the list
+`[1; 2; 3]` models the queue obtained by pushing `1`, `2`, `3` in this order.
+Note that this proof relies entirely on the theorems from the specification:
+the proof would still work even if we completely changed the implementation
+and the definition of `models` (in this sense, it is an implementation detail!).
+
+<div class="tuto">
+```alectryon
+Example example1 : models [1; 2; 3] (push 3 (push 2 (push 1 empty))).
+Proof.
+  pose proof (models_empty (a := nat)) as H.
+  apply (models_push 1) in H. cbn [ app ] in H.
+  apply (models_push 2) in H. cbn [ app ] in H.
+  apply (models_push 3) in H. cbn [ app ] in H.
+  apply H.
+Qed.
+```
+</div>
+
+The above proof only uses `push` and `empty`.
+Here is another example, proving that `pop` on the above queue produces the
+first element `1`.
+
+<div class="tuto">
+```alectryon
+Example example2 :
+  match pop (push 3 (push 2 (push 1 empty))) with
+  | None => False
+  | Some (y, _) => y = 1
+  end.
+Proof.
+  pose proof example1 as H.
+  apply models_pop in H.
+  destruct H as [q' [Hpop Hq']].
+  rewrite Hpop.
+  reflexivity.
+Qed.
+```
+</div>
+
+Both the implementation and the specification may contain mistakes.
+Formal verification, by connecting the two, weeds out "mismatches" between
+them.
+Of course, matching mistakes might happen on both sides. This motivates
+unit tests as we do above, and other forms of *metaspecification*
+(specification of the specification).
```

Comment:
It might be useful to have some examples of metaspecifications. For example, something like `models (x : xs) p -> forall y, fst <$> pop (push y p) = fst <$> pop p`?